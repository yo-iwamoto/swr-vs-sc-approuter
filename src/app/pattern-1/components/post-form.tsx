"use client";

import { useCreatePostMutation } from "@/app/pattern-1/mutations/use-create-post-mutation";
import type { FormEvent } from "react";
import { Button, FormControl, Stack, Textarea } from "smarthr-ui";

export function PostForm() {
  const mutation = useCreatePostMutation();

  const onSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    const form = e.currentTarget;
    const content = new FormData(form).get("content");
    if (typeof content !== "string") return;

    await mutation.trigger({ content });
    form.reset();
  };

  return (
    <section className="max-w-lg mx-auto w-full">
      <form onSubmit={onSubmit}>
        <Stack className="items-center gap-1">
          <h1 className="text-lg font-bold">新規ポスト</h1>

          <FormControl title="ポストの本文" className="w-full">
            <Textarea
              name="content"
              required
              maxLength={140}
              className="w-full py-2"
              data-1p-ignore
              rows={5}
            />
          </FormControl>

          <Button
            type="submit"
            variant="primary"
            className="w-full overflow-hidden h-[42px]"
            loading={mutation.isMutating}
          >
            投稿
          </Button>
        </Stack>
      </form>
    </section>
  );
}
