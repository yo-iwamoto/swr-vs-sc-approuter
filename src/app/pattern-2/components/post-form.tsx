import { FormControl, Stack, Textarea } from "@/components/client-ui";
import { api } from "@/lib/api";
import { callApi } from "../server/call-api";
import { revalidateTag } from "next/cache";
import { FormStatusButton } from "./form-status-button";

export function PostForm() {
  const createPostAction = async (data: FormData) => {
    "use server";

    const content = data.get("content");
    if (typeof content !== "string") return;

    await callApi(api.posts.$url().toString(), {
      method: "POST",
      body: JSON.stringify({
        content,
      }),
    });
    revalidateTag("posts");
  };

  return (
    <section className="max-w-lg mx-auto w-full">
      <form action={createPostAction}>
        <Stack className="items-center gap-1">
          <h1 className="text-lg font-bold">新規ポスト</h1>

          <FormControl title="ポストの本文" className="w-full">
            <Textarea
              name="content"
              required
              maxLength={140}
              className="w-full py-2"
              data-1p-ignore
              rows={5}
            />
          </FormControl>

          <FormStatusButton
            type="submit"
            variant="primary"
            className="w-full overflow-hidden h-[42px]"
          >
            投稿
          </FormStatusButton>
        </Stack>
      </form>
    </section>
  );
}
