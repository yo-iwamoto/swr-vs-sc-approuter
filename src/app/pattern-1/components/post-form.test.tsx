import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { PostForm } from "./post-form";

const triggerMock = vi.fn();
vi.mock("@/app/pattern-1/mutations/use-create-post-mutation", () => ({
  useCreatePostMutation: vi.fn(() => ({
    trigger: triggerMock,
  })),
}));

const mutateMock = vi.hoisted(() => vi.fn());
vi.mock("swr", () => ({
  mutate: mutateMock,
}));

describe("PostForm", () => {
  it("サブミットすると、ポストの作成リクエストが行われること", async () => {
    render(<PostForm />);

    await userEvent.type(
      screen.getByRole("textbox", { name: "ポストの本文" }),
      "Hello world!",
    );
    await userEvent.click(screen.getByRole("button", { name: "投稿" }));

    expect(triggerMock).toHaveBeenLastCalledWith({ content: "Hello world!" });
  });

  it("サブミット後、入力内容がリセットされること", async () => {
    render(<PostForm />);

    await userEvent.type(
      screen.getByRole("textbox", { name: "ポストの本文" }),
      "Hello world!",
    );
    await userEvent.click(screen.getByRole("button", { name: "投稿" }));

    expect(screen.getByRole("textbox", { name: "ポストの本文" })).toHaveValue(
      "",
    );
  });

  it("サブミット後、poss が mutate されること", async () => {
    render(<PostForm />);

    await userEvent.type(
      screen.getByRole("textbox", { name: "ポストの本文" }),
      "Hello world!",
    );
    await userEvent.click(screen.getByRole("button", { name: "投稿" }));

    expect(mutateMock).toHaveBeenLastCalledWith(["posts"]);
  });

  it("snapshot", () => {
    const { asFragment } = render(<PostForm />);

    expect(asFragment()).toMatchSnapshot();
  });
});
