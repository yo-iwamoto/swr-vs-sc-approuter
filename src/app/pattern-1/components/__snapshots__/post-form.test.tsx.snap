// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostForm > snapshot 1`] = `
<DocumentFragment>
  <section
    class="max-w-lg mx-auto w-full"
  >
    <form>
      <div
        class="shr-flex-col shr-justify-start [&_>_*]:shr-my-0 shr-flex shr-space-y-1 items-center gap-1"
      >
        <h1
          class="text-lg font-bold"
        >
          新規ポスト
        </h1>
        <div
          class="shr-flex-col shr-justify-start [&_>_*]:shr-my-0 shr-flex shr-space-y-0.5 sc-ilxebA fSuyNt w-full smarthr-ui-FormGroup"
        >
          <label
            class="shr-flex-wrap [&:empty]:shr-gap-0 shr-flex shr-gap-y-0.5 shr-gap-x-0.5 shr-items-center sc-jIBlqr hndDbA smarthr-ui-FormGroup-label"
            for=":r4:"
            id=":r5:"
          >
            <span
              class="shr-text-base shr-font-bold"
            >
              ポストの本文
            </span>
          </label>
          <div
            class="sc-gYrqIg keLuCv"
          >
            <textarea
              class="smarthr-ui-Textarea-textarea shr-border-shorthand shr-my-[unset] shr-box-border shr-rounded-m shr-bg-white shr-p-0.5 shr-text-base shr-leading-normal shr-text-black shr-opacity-100 contrast-more:shr-border-high-contrast placeholder:shr-text-grey focus-visible:shr-focus-indicator disabled:shr-pointer-events-none disabled:shr-bg-column disabled:shr-text-disabled disabled:placeholder:shr-text-disabled shr-border-default w-full py-2"
              data-1p-ignore="true"
              id=":r4:"
              name="content"
              required=""
              rows="5"
            />
            <span
              class="smarthr-ui-Textarea-counter shr-block shr-text-sm"
            >
              あと
              <span
                class="shr-font-bold shr-text-grey"
              >
                140
              </span>
              文字
            </span>
          </div>
        </div>
        <button
          class="disabled:shr-cursor-not-allowed disabled:shr-bg-clip-padding [&_.smarthr-ui-Icon]:forced-colors:disabled:shr-fill-[GrayText] shr-box-border shr-cursor-pointer shr-inline-flex shr-justify-center shr-items-center shr-gap-0.5 shr-text-center shr-whitespace-nowrap shr-rounded-m shr-border-shorthand shr-font-inherit shr-font-bold shr-leading-none focus-visible:shr-focus-indicator contrast-more:shr-border-high-contrast [&_.smarthr-ui-Icon]:shr-block [&_svg]:shr-block shr-text-base shr-px-1 shr-py-0.75 shr-border-main shr-bg-main shr-text-white focus-visible:shr-border-main-darken focus-visible:shr-bg-main-darken hover:shr-border-main-darken hover:shr-bg-main-darken disabled:shr-border-main/50 disabled:shr-bg-main/50 disabled:shr-text-white/50 smarthr-ui-Button w-full overflow-hidden h-[42px]"
          type="submit"
        >
          <span
            class="shr-min-w-0"
          >
            投稿
          </span>
        </button>
      </div>
    </form>
  </section>
</DocumentFragment>
`;
